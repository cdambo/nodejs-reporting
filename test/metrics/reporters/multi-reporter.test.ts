import { invoke, map } from "lodash";
import StatsDReporter from "../../../src/metrics/reporters/statsd-reporter";
import MultiReporter from "../../../src/metrics/reporters/multi-reporter";
import InMemoryReporter from "../../../src/metrics/reporters/in-memory-reporter";
import ConsoleReporter from "../../../src/metrics/reporters/console-reporter";
import MetricsReporter from "../../../src/metrics/reporters/metrics-reporter";

const getReporter = (): MultiReporter =>
  new MultiReporter({
    reporters: [
      new StatsDReporter(),
      new InMemoryReporter(),
      new ConsoleReporter()
    ]
  });

const reportsOf = (
  method: keyof MetricsReporter,
  ...args: unknown[]
): ReturnType<typeof jest.spyOn>[] => {
  const multiReporter = getReporter();
  const mocks = map(
    multiReporter.reporters,
    (reporter: MetricsReporter): ReturnType<typeof jest.spyOn> =>
      jest.spyOn(reporter, method).mockImplementation()
  );
  invoke(multiReporter, method, ...args);
  return map(
    mocks,
    (mock: ReturnType<typeof jest.spyOn>): ReturnType<typeof jest.spyOn> =>
      mock.mock.calls
  );
};

describe("MultiReporter", (): void => {
  describe("Constructor", (): void => {
    it("Creates a MultiReporter", (): void => {
      const reporter = new MultiReporter({
        reporters: [
          new StatsDReporter(),
          new InMemoryReporter(),
          new ConsoleReporter()
        ]
      });
      expect(reporter).toMatchSnapshot();
    });
  });

  describe("timing", (): void => {
    it("Reports timing metric", (): void => {
      expect(
        reportsOf("timing", "timing_test", 20, 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("increment", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("increment", "increment_test", 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("incrementBy", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("incrementBy", "increment_by_test", 10, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("decrement", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("decrement", "decrement_test", 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("decrementBy", (): void => {
    it("Reports counter metric", (): void => {
      expect(
        reportsOf("decrementBy", "decrement_by_test", 10, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("gauge", (): void => {
    it("Reports gauge metric", (): void => {
      expect(
        reportsOf("gauge", "gauge_test", 10, 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });

  describe("histogram", (): void => {
    it("Reports histogram metric", (): void => {
      expect(
        reportsOf("histogram", "histogram_test", 20, 0.5, {
          tag1: "one",
          tag2: "two"
        })
      ).toMatchSnapshot();
    });
  });
});
