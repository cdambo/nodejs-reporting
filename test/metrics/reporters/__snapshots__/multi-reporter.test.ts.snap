// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiReporter Constructor Creates a MultiReporter 1`] = `
MultiReporter {
  "reporters": Array [
    StatsDReporter {
      "client": Client {
        "ephemeral_socket": null,
        "global_tags": undefined,
        "host": "localhost",
        "last_used_timer": null,
        "port": 8125,
        "socket": null,
      },
      "context": Context {
        "globalTags": Object {},
      },
      "format": Object {
        "format": [Function],
      },
    },
    InMemoryReporter {
      "context": Context {
        "globalTags": Object {},
      },
      "format": Object {
        "format": [Function],
      },
      "reports": Array [],
    },
    ConsoleReporter {
      "context": Context {
        "globalTags": Object {},
      },
      "format": Object {
        "format": [Function],
      },
      "output": [Function],
    },
  ],
}
`;

exports[`MultiReporter decrement Reports counter metric 1`] = `
Array [
  Array [
    Array [
      "decrement_test",
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "decrement_test",
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "decrement_test",
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;

exports[`MultiReporter decrementBy Reports counter metric 1`] = `
Array [
  Array [
    Array [
      "decrement_by_test",
      10,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "decrement_by_test",
      10,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "decrement_by_test",
      10,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;

exports[`MultiReporter gauge Reports gauge metric 1`] = `
Array [
  Array [
    Array [
      "gauge_test",
      10,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "gauge_test",
      10,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "gauge_test",
      10,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;

exports[`MultiReporter histogram Reports histogram metric 1`] = `
Array [
  Array [
    Array [
      "histogram_test",
      20,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "histogram_test",
      20,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "histogram_test",
      20,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;

exports[`MultiReporter increment Reports counter metric 1`] = `
Array [
  Array [
    Array [
      "increment_test",
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "increment_test",
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "increment_test",
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;

exports[`MultiReporter incrementBy Reports counter metric 1`] = `
Array [
  Array [
    Array [
      "increment_by_test",
      10,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "increment_by_test",
      10,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "increment_by_test",
      10,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;

exports[`MultiReporter timing Reports timing metric 1`] = `
Array [
  Array [
    Array [
      "timing_test",
      20,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "timing_test",
      20,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
  Array [
    Array [
      "timing_test",
      20,
      0.5,
      Object {
        "tag1": "one",
        "tag2": "two",
      },
    ],
  ],
]
`;
